### PR TITLE
tinytqdm write support

### PR DESCRIPTION
**Pull request by francislata**
**Overview**
This PR adds support for `write` that is commonly used in quite a few [examples](https://github.com/tinygrad/tinygrad/blob/master/examples/mlperf/model_train.py#L230).

Here's an example of it after running the following:
```python
from tinytqdm.helpers import tqdm
for i in tqdm(range(10)):
  tqdm.write(str(i))
```
Screenshot 2024-09-04 at 15 23 14

### CODE REVIEW

File: test/unit/test_tqdm.py [hunk 1]
```diff
@@ -209,6 +207,20 @@ def test_tqdm_output_custom_nolen_total(self, mock_terminal_size, mock_stderr):
...
+    ncols, tqdm_fp = random.randint(80, 120), StringIO()
+    mock_terminal_size.return_value = namedtuple(field_names='columns', typename='terminal_size')(ncols)
+    mock_stderr.truncate(0)
+    tqdm_fp.truncate(0)
+    for i in tinytqdm(range(10)):
+      time.sleep(0.01)
+      tinytqdm.write(str(i))
+      tqdm.write(str(i), file=tqdm_fp)
+      tinytqdm_out, tqdm_out = mock_stderr.getvalue().split("\r\033[K")[-1], tqdm_fp.getvalue().split(f"{i-1}\n")[-1]
+      self.assertEqual(tinytqdm_out, tqdm_out)
```
Comment (chenyuxyz) [inline test/unit/test_tqdm.py:222]: can you also test that `0, 1, 2, 3, ...` are in the output?

**Review by chenyuxyz — COMMENTED** [review]
looks good, minor comment about test. Also have you run any full mlperf example? running anyone suffices but want to make sure the output logs are fine

**Comment by francislata** [comment]
> looks good, minor comment about test. Also have you run any full mlperf example? running anyone suffices but want to make sure the output logs are fine

let me test it with a full mlperf example and update the test case as well .

**Comment by francislata** [comment]
@chenyuxyz 
> Also have you run any full mlperf example? running anyone suffices but want to make sure the output logs are fine

the outputs do match between `tqdm` and `tinytqdm`:

|   **tinytqdm**   |   **tqdm**   |
| --------------- | --------------- |
| tinytqdm | tqdm |

**Comment by chenyuxyz** [comment]
https://github.com/tinygrad/tinygrad/actions/runs/10963049261/job/30443684158?pr=6359
it seems to have one more newline compared to tqdm (you can check the ones in master)
image

**Comment by francislata** [comment]
> https://github.com/tinygrad/tinygrad/actions/runs/10963049261/job/30443684158?pr=6359 it seems to have one more newline compared to tqdm (you can check the ones in master)

@chenyuxyz that's interesting. running the same command locally, i didn't find anything unusual on my terminal printing newlines. i added a force flush when `write` is called. also, i was going to verify this myself but since i don't have the access to push to `update_benchmarks`, i'll leave it up to you to check again. i'll try to reproduce it on my end, if possible.

**Comment by chenyuxyz** [comment]
just pushed, still seeing extra rows

**Comment by francislata** [comment]
@chenyuxyz - it looks like terminal and [wandb](https://wandb.ai/flata/tinygrad-examples_mlperf/runs/bez0ljbh/logs) looks good:
Screenshot 2024-10-16 at 13 53 29
Screenshot 2024-10-16 at 13 52 42

**Comment by chenyuxyz** [comment]
training bert seems fine too `https://wandb.ai/chenyuxyz/MLPerf-BERT/runs/m086t2l8?nw=nwuserchenyuxyz`. will merge after CI finishes

**Pull request merged**